ns for fx: add support for logging inputs

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53828 ns for fx: move shadow lstm test to new API
* #53819 ns for fx: move shadow activations linear test to new API
* #53818 ns for fx: move shadow activations conv test to new API
* #53779 ns for fx: add support for lstm activation matching
* #53777 ns for fx: move linear activation test case to new API
* #53776 ns for fx: move compare activations for conv test to new API
* #53772 ns for fx: move lstm dynamic weight test case to new API
* #53765 ns for fx: move linear dynamic weight test case to new API
* #53764 ns for fx: move linear weight test case to new API
* #53761 ns for fx: move conv weight test case to new API
* #53748 ns for fx: extract common code in tests to util functions
* #53729 ns for fx: clean up API naming
* **#53249 ns for fx: add support for logging inputs**
* #53248 ns for fx: change dtype cast from once per N node to once per node
* #53075 ns for fx: change API to take nn.Module instead of GraphModule

Summary:

This PR implements the option to log inputs for FX Numeric Suite.  The user facing api looks like

```
def prepare_model_outputs(..., should_log_inputs : bool = False)
def prepare_model_with_stubs(..., should_log_inputs : bool = False)
```

The output data now looks like

```
{
  "layer1": {
    "node_inputs": {
      "model1": [{
        "values": ...,
        ...,
      }],
    },
    "node_outputs": {
      ...,
    }
  },
  ...  // other layers
}
```

One key design decision taken here is that an input logger logs the output of previous nodes, instead of logging the input of the current node.  This matters for a signature such as `cat([x1, x2, x3])`.  We are inserting three input loggers here (for x1, x2, and x3), instead of a single input logger for `[x1, x2, x3]`.  This was chosen in order to preserve the structure of the original graph as much as possible and keep flexibility for future optimizations.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931225](https://our.internmc.facebook.com/intern/diff/D26931225)

vkuzo committed:
ns for fx: add support for logging inputs

Summary:

TODO: fill out

Test Plan:

TODO: fill out

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
save intermediate work on "ns for fx: add support for logging inputs"


Summary:

TODO: fill out

Test Plan:

TODO: fill out

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for logging inputs"


Summary:

TODO: fill out

Test Plan:

TODO: fill out

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for logging inputs"


Summary:

This PR implements the option to log inputs for FX Numeric Suite.  The user facing api looks like

```
def prepare_model_outputs(..., should_log_inputs : bool = False)
def prepare_model_with_stubs(..., should_log_inputs : bool = False)
```

The output data looks like

```
{
  // inputs to layer1
  "layer1_in": {
    "model1": {
      "type": "node_input",
      ...
    },
  },
  // outputs of layer1
  "layer1": {
    "model1": {
      "type": "node_output",
      ...
    },
  },
  ...
}
```

One key design decision taken here is that an input logger logs the output of previous nodes, instead of logging the input of the current node.  This matters for a signature such as `cat([x1, x2, x3])`.  We are inserting three input loggers here (for x1, x2, and x3), instead of a single input logger for `[x1, x2, x3]`.  This was chosen in order to preserve the structure of the original graph as much as possible and keep flexibility for future optimizations.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931225](https://our.internmc.facebook.com/intern/diff/D26931225)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for logging inputs"


Summary:

This PR implements the option to log inputs for FX Numeric Suite.  The user facing api looks like

```
def prepare_model_outputs(..., should_log_inputs : bool = False)
def prepare_model_with_stubs(..., should_log_inputs : bool = False)
```

The output data now looks like

```
{
  "layer1": {
    "node_inputs": {
      "model1": [{
        "values": ...,
        ...,
      }],
    },
    "node_outputs": {
      ...,
    }
  },
  ...  // other layers
}
```

One key design decision taken here is that an input logger logs the output of previous nodes, instead of logging the input of the current node.  This matters for a signature such as `cat([x1, x2, x3])`.  We are inserting three input loggers here (for x1, x2, and x3), instead of a single input logger for `[x1, x2, x3]`.  This was chosen in order to preserve the structure of the original graph as much as possible and keep flexibility for future optimizations.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931225](https://our.internmc.facebook.com/intern/diff/D26931225)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for logging inputs"


Summary:

This PR implements the option to log inputs for FX Numeric Suite.  The user facing api looks like

```
def prepare_model_outputs(..., should_log_inputs : bool = False)
def prepare_model_with_stubs(..., should_log_inputs : bool = False)
```

The output data now looks like

```
{
  "layer1": {
    "node_inputs": {
      "model1": [{
        "values": ...,
        ...,
      }],
    },
    "node_outputs": {
      ...,
    }
  },
  ...  // other layers
}
```

One key design decision taken here is that an input logger logs the output of previous nodes, instead of logging the input of the current node.  This matters for a signature such as `cat([x1, x2, x3])`.  We are inserting three input loggers here (for x1, x2, and x3), instead of a single input logger for `[x1, x2, x3]`.  This was chosen in order to preserve the structure of the original graph as much as possible and keep flexibility for future optimizations.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931225](https://our.internmc.facebook.com/intern/diff/D26931225)

[ghstack-poisoned]

hx89 commented on torch/quantization/ns/numeric_suite_core_apis_fx.py:
```diff
@@ -28,11 +27,69 @@
-    WEIGHT = 'weight'
-    NODE_OUTPUT = 'node_output'
+
+class OutputLogger(nn.Module):
```
Maybe we can think of another name for this logger later, since we also use it to log input and current name might be a little confusing.

vkuzo committed:
Update on "ns for fx: add support for logging inputs"


Summary:

This PR implements the option to log inputs for FX Numeric Suite.  The user facing api looks like

```
def prepare_model_outputs(..., should_log_inputs : bool = False)
def prepare_model_with_stubs(..., should_log_inputs : bool = False)
```

The output data now looks like

```
{
  "layer1": {
    "node_inputs": {
      "model1": [{
        "values": ...,
        ...,
      }],
    },
    "node_outputs": {
      ...,
    }
  },
  ...  // other layers
}
```

One key design decision taken here is that an input logger logs the output of previous nodes, instead of logging the input of the current node.  This matters for a signature such as `cat([x1, x2, x3])`.  We are inserting three input loggers here (for x1, x2, and x3), instead of a single input logger for `[x1, x2, x3]`.  This was chosen in order to preserve the structure of the original graph as much as possible and keep flexibility for future optimizations.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931225](https://our.internmc.facebook.com/intern/diff/D26931225)

[ghstack-poisoned]

vkuzo commented:
note: this was merged, not sure why it doesn't show up here
